### PR TITLE
[Blazor] Prevent error boundaries from handling exceptions of type `NavigationException`

### LINKED ISSUE #52777: Error boundaries may block redirects by handling exceptions of type `NavigationException`
When you try to navigate to another page by using the NavigationManager.NavigateTo in a blazor web app introduced in .net8,the following  exception is thrown : 
**Microsoft.AspNetCore.Components.NavigationException: 'Exception of type 'Microsoft.AspNetCore.Components.NavigationException' was thrown.'**
Even if you put the logic in  a try/catch block,the exception would be still thrown and the page would  stay on the current route.This means that if,for example,you wanted to create a login page that should take the user to the main page after a successful login form submission,the user would still stay on the login page!

Even though the issue had previously been brought up by @danroth27 at #49143 and closed as done later on,the problem still persists in the final version of .net8.I've also  created a sample  project that you can check out in the following repo:
[Blazor Web App With Custom Cookie-based Authentication](https://github.com/meisam-dehghan/BlazorWebApp)

### PR DESCRIPTION
**Pull request by MackinnonBuck**
# Prevent error boundaries from handling exceptions of type `NavigationException`

Fixes an issue where error boundaries may handle `NavigationException` instances and treat them as uncaught errors rather than signals to redirect.

## Description

This PR fixes the issue by ignoring error boundaries if the exception getting dispatched is a `NavigatonException`.

Fixes #52777

### CODE REVIEW

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 1]
```diff
@@ -1066,6 +1066,15 @@ private void HandleExceptionViaErrorBoundary(Exception error, ComponentState? er
...
 
+        // We don't allow NavigationException instances to be caught by error boundaries.
+        // These are special exceptions whose purpose is to be as invisible as possible to
+        // user code and bubble all the way up to get handled by the framework as a redirect.
+        if (error is NavigationException)
+        {
+            HandleException(error);
+            return;
+        }
+
```
Comment (MackinnonBuck) [inline src/Components/Components/src/RenderTree/Renderer.cs:1077]: There are some things to consider about this approach:
1. There could arguably be cases where `NavigationException` _should_ be treated as an error (e.g., if it's thrown during interactive rendering). It might be preferable to allow error boundaries to handle `NavigationException`s during interactive rendering, but not static server rendering. To support this, we could alternatively update the `ErrorBoundaryBase` implementation to ignore `NavigationException` instances when prerendering.
2. Since `NavigationException` is not `sealed`, someone could extend it to create more exceptions that bypass error boundaries. I'm not totally sure why someone would do this, but if we see it as a potential antipattern and want to proactively discourage it, we could update this to:
    ```csharp
    if (error.GetType() == typeof(NavigationException))
    {
        HandleException(error);
        return;
    }
    ```

Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/Renderer.cs:1077]: 1. It's good to think this through, but TBH it's not clear to me that we need to introduce further distinctions between interactive and static rendering on this behavior. Unless there's a known use case, that is.
2. That's a great thing to anticipate. I think the only reason for subclassing `NavigationException` is if you're also implementing a custom rendering platform, and you want a way to extend the meaning of "navigation" on your platform with other fields/flags (and in that case, you don't want it to trigger an error boundary - it would just be an obscure bug you probably didn't realise you caused). People can't really subclass it for arbitrary non-navigation-related reasons because the framework is still going to catch their subclass and use it to perform a navigation, hence it seems unlikely the feature could be misused.

So in summary my guess is your existing design/implementation is already the best design. Unless you know of more reasons we should anticipate misuse/problems.

Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/Renderer.cs:1077]: > I think the only reason for subclassing NavigationException is if you're also implementing a custom rendering platform

Actually even that use case doesn't really make sense to me now I think about it. It would have to be a new custom SSR rendering platform, not an interactive one. And who's doing that?

So if we wanted to be super risk-averse, we could have logic like:

```cs
if (error is NavigationException)
{
  if (error.GetType() == typeof(NavigationException))
  {
      HandleException(error);
      return;
  }
  else
  {
      throw new InvalidOperationException("We don't know what you're trying to do, but stop doing it");
  }
}
```

Comment (MackinnonBuck) [reply src/Components/Components/src/RenderTree/Renderer.cs:1077]: > So in summary my guess is your existing design/implementation is already the best design. Unless you know of more reasons we should anticipate misuse/problems.

Yeah, I agree - I don't anticipate that either of the points I brought up will become real problems, but I figured it might be a good idea to bring them up in case anyone thought otherwise. There's room to revise this in the future if we feel the need to as well.

**Review by javiercn — APPROVED** [review]
